keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1504 @@
+<<[/UNRESOLVED]>>
+```
+
+#### `NodePrepareResource`
+
+This RPC is called by kubelet when a Pod that wants to use the
+specified resource is scheduled on a node.  The Plugin SHALL assume
+that this RPC will be executed on the node where the resource will be
+used.  The Plugin SHALL return device name and kind for allocated
+device[s].
+
+The Plugin SHALL create or update json file[s] in CDI format for each
+allocated device. These files SHALL be used by runtime to update
+runtime configuration before creating containers that use the
+device[s].
+
+This operation SHALL do as little work as possible as it’s called
+after a pod is scheduled to a node. All potentially failing operations
+SHALL be done during allocation phase.
+
+This operation MUST be idempotent. If the resource corresponding to
+the `resource_id` has already been prepared, the Plugin MUST reply `0
+OK`.
+
+If this RPC failed, or kubelet does not know if it failed or not, it
+MAY choose to call `NodePrepareResource` again, or choose to call
+`NodeUnprepareResource`.
+
+After a successful call, Kubelet MUST pass device names and kinds to
+the runtime through the CRI protocol.
+
+```
+<<[UNRESOLVED @bart0sh]>>
+CRI protocol may need to be extended for
+this purpose, e.g. device id can be added to the CRI Device structure.
+<<[/UNRESOLVED]>>
+```
+
+```protobuf
+message NodePrepareResourceRequest {
+  // The UID of the ResourceClaim. This field is REQUIRED.
+  string resource_uid = 1;
+}
+
+message NodePrepareResourceResponse {
+  // These are the additional devices that kubelet must
+  // make available via the container runtime. A resource
+  // may have zero or more devices.
+  repeated CDIDevice devices = 1;
+}
+
+message CDIDevice {
+  // Kind is the string that together with the name identifies a device
+  // (https://github.com/container-orchestrated-devices/container-device-interface/blob/master/SPEC.md#kind).
+  string kind = 1;
+  // Name is the name that within its kind uniquely identifies a
+  // device (https://github.com/container-orchestrated-devices/container-device-interface/blob/master/SPEC.md#cdi-devices).
+  string name = 2;
+}
+```
+
+##### NodePrepareResource Errors
+
+If the plugin is unable to complete the NodePrepareResource call
+successfully, it MUST return a non-ok gRPC code in the gRPC status.
+If the conditions defined below are encountered, the plugin MUST
+return the specified gRPC error code.  Kublet MUST implement the
+specified error recovery behavior when it encounters the gRPC error
+code.
+
+| Condition | gRPC Code | Description | Recovery Behavior |
+|-----------|-----------|-------------|-------------------|
+| Resource does not exist | 5 NOT_FOUND | Indicates that a resource corresponding to the specified `resource_id` does not exist. | Caller MUST verify that the `resource_id` is correct and that the resource is accessible and has not been deleted before retrying with exponential back off. |
+ 
+
+#### `NodeUnprepareResource`
```

Comment:
Similar to my comment above, I think we should rename `NodeUnprepareResource` to `NodeDeallocateResource` or simply `DeallocateResource`.

Reply:
See https://github.com/kubernetes/enhancements/pull/3064#discussion_r770858491: because the "real" allocation happens earlier, these node-local calls are really just about "preparing" the resource for usage by a pod.